src/main/java/com/stuypulse/robot/commands/swerve/SwerveDriveBalanceWithDualPID.java

```diff
@@ -0,0 +1,131 @@
+
+        DISTANCE_THRESHOLD = AutoEngage.DISTANCE_THRESHOLD.doubleValue();
+        ANGLE_THRESHOLD = AutoEngage.ANGLE_THRESHOLD.doubleValue();
+
+        this.swerve = SwerveDrive.getInstance();
+        this.odometry = Odometry.getInstance();
+
+        tiltController = new PIDController(Tilt.P, Tilt.I, Tilt.D);
+        translationController = new PIDController(Translation.P, Translation.I, Translation.D);
+
+        gyroController = new PIDController(Constants.kP, Constants.kI, Constants.kD);
+
+        addRequirements(swerve);
+    }
+
+    @Override
+    public void initialize() {
+        odometry.overrideNoise(true);
+    }
+
+    private Rotation2d getBalanceAngle() {
+        Rotation2d pitch = swerve.getGyroPitch();
+        Rotation2d roll = swerve.getGyroRoll();
+        Rotation2d yaw = odometry.getRotation();
+        
+        double facingSlope = pitch.getTan() * yaw.getCos() + roll.getTan() * yaw.getSin();
+        double maxSlope = Math.sqrt(Math.pow(roll.getTan(), 2) + Math.pow(pitch.getTan(), 2));
+
+        return Rotation2d.fromRadians(Math.signum(facingSlope) * Math.atan(maxSlope));
+    }
+
+    @Override
+    public void execute() {
+        double balanceAngle = getBalanceAngle().getDegrees();
+        double target = Units.inchesToMeters(CHARGING_STATION_CENTER.get());
+        double offset = tiltController.update(0, balanceAngle) / Math.cos(Math.toRadians(balanceAngle));
+
+        target += offset;
+        double currentPosition = odometry.getTranslation().getX();
+
+        double velocity;
+
+        if (translationController.isDone(AutoEngage.DISTANCE_THRESHOLD.getAsDouble())) {
+            velocity = gyroController.update(0, -1 * balanceAngle);
+            swerve.setChassisSpeeds(ChassisSpeeds.fromFieldRelativeSpeeds(
+                                            new ChassisSpeeds(velocity, 0.0, 0.0), 
+                                            odometry.getRotation()));
+        } else {
+            velocity = translationController.update(target, currentPosition);
+
+            swerve.setChassisSpeeds(ChassisSpeeds.fromFieldRelativeSpeeds(
+                                            new ChassisSpeeds(velocity, 0.0, 0.0), 
+                                            odometry.getRotation()));
+            
+            SmartDashboard.putNumber("Auto Engage/Balance Angle", balanceAngle);
+            SmartDashboard.putNumber("Auto Engage/Velocity", velocity);
+            SmartDashboard.putNumber("Auto Engage/Target", target);
+        }
+    }
+
+    @Override 
+    public boolean isFinished() {
+        return gyroController.isDone(AutoEngage.ANGLE_THRESHOLD.getAsDouble());
+    }
+
+    @Override 
+    public void end(boolean interrupted) {
+        swerve.stop();
+        odometry.overrideNoise(false);
+        
+        Plant.getInstance().engage();
+    }
+
+    public Command pointWheels() {
+        return andThen(new SwerveDrivePointWheels(Rotation2d.fromDegrees(90)));
```

Comment:
Should we just have this typed out in robot container or in a sequential command group? You can resolve this if you want to keep this anyway (just rename it thenPointWheels if you do)